sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -47,6 +48,34 @@ trait Predicate extends Expression {
   override def dataType: DataType = BooleanType
 }
 
+object Predicate extends PredicateHelper {
+  def toCNF(predicate: Expression, maybeThreshold: Option[Double] = None): Expression = {
+    val cnf = new CNFExecutor(predicate).execute(predicate)
+    val threshold = maybeThreshold.map(predicate.size * _).getOrElse(Double.MaxValue)
+    if (cnf.size > threshold) predicate else cnf
```

Comment:
When the threshold is exceeded, the original predicate rather than the intermediate converted predicate is returned. This because the intermediate result may not be in CNF, thus:

1. It doesn't bring much benefit for filter push-down, and
2. It's much larger than the original predicate and brings extra evaluation cost.